TeamCode/src/main/java/org/firstinspires/ftc/teamcode/ultimategoal/vision/GoalFinder.java

```diff
@@ -0,0 +1,165 @@
+package org.firstinspires.ftc.teamcode.ultimategoal.vision;
+
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.MatrixUtils;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.linear.RealVector;
+import org.opencv.core.*;
+import org.opencv.imgproc.Imgproc;
+import org.opencv.imgproc.Moments;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.opencv.imgproc.Imgproc.*;
+
+public class GoalFinder {
+
+    static final double[][] CAMERA_MATRIX = {
+            {1468, 0, 0},
+            {0, 1468, 0},
+            {960, 540, 1}
+    };
+
+    static final Scalar ORANGE = new Scalar(0, 165, 255);
+    static final Scalar BLACK = new Scalar(0, 0, 0);
+
+    public static class GoalLocationData {
+        double yaw;
+        double pitch;
+        double x;
+        double y;
+
+        public GoalLocationData(double yaw, double pitch, double x, double y) {
+            this.yaw = yaw;
+            this.pitch = pitch;
+            this.x = x;
+            this.y = y;
+        }
+
+        public double getYaw() {
+            return yaw;
+        }
+
+        public double getPitch() {
+            return pitch;
+        }
+
+        public double getX() {
+            return x;
+        }
+
+        public double getY() {
+            return y;
+        }
+
+        public String toString() {
+            return String.format("Angle (degrees, euler xy): {%f, %f}, Location (in pixels): {%f, %f}", getYaw(), getPitch(), getX(), getY());
+        }
+    }
+
+    /**
+     * Finds the largest contour by enclosed area in a list
+     *
+     * @param contours
+     * @return the index of contours which the largest is located
+     */
+    static int largestContour(final List<MatOfPoint> contours) {
+        if (contours.size() == 0)
+            return -1;
+
+        double maxVal = 0;
+        int maxValIdx = 0;
+        for (int contourIdx = 0; contourIdx < contours.size(); contourIdx++) {
+            double contourArea = Imgproc.contourArea(contours.get(contourIdx));
+            if (maxVal < contourArea) {
+                maxVal = contourArea;
+                maxValIdx = contourIdx;
+            }
+        }
+
+        return maxValIdx;
+    }
+
+    /**
+     * This method takes two points, point1 and point2, and then projects these points to form two rays. It then returns the angle between these two rays
+     * in radians.
```

Comment:
Doesn't the distance from these points affect the angle between them? How does it project these points to form rays? Not sure I quite understand the intuition.

Reply:
This method ended up being overkill, however I left it in in case I find that it works better. The first thing is that I'm using a 3x3 camera matrix, since the last column is all zeroes, and for the operations I'm performing that works fine. Let x be some 3D coordinate, y being some 2D location on the image, and K is our camera matrix. y~Kx.

Pinhole cameras basically can be mapped (projected) using the equation **y**=(f/x3)(**x**). f is the focal length of the camera, y is our 2d coordinate, and x is the 2d components of our 3d coordinate. x3 is the final component. We convert it to homogenous coordinates (like x,y,z), so for y we just use (y1,y2,1) (i'm using y as a vector, y1 and y2 are the x and y components of it). Instead of equality, we use ~ for proportionality, which gives us our matrix. 

This is overkill, and also just a pain, since we need to calibrate the camera and such -- my current solution of just dividing up the field of view is working fine, and I left it in there just for future reference. Also, the matrix I used for the camera has some funny values that aren't the resolution -- that's just because I pre-applied the focal length